src/types/type_param.rs

```diff
@@ -0,0 +1,61 @@
+//! Type Parameters
+//!
+//! Parameters for [`TypeDef`]s provided by extensions
+//!
+//! [`TypeDef`]: crate::resource::TypeDef
+
+use crate::ops::constant::HugrIntValueStore;
+
+use super::{ClassicType, SimpleType};
+
+/// A parameter declared by an OpDef. Specifies a value
+/// that must be provided by each operation node.
+/// TODO any other 'leaf' types? We specifically do not want float.
+/// bool should eventually be a Sum type (Container).
+#[derive(Clone, Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+pub enum TypeParam {
+    /// Argument is a [TypeArg::Type] - classic or linear
+    Type,
+    /// Argument is a [TypeArg::ClassicType]
+    ClassicType,
+    /// Argument is an integer
+    Int,
+    /// Node must provide a [TypeArg::List] (of whatever length)
+    /// TODO it'd be better to use [`Container`] here.
+    ///
+    /// [`Container`]: crate::types::simple::Container
+    List(Box<TypeParam>),
+}
+
+/// A statically-known argument value to an operation.
+#[derive(Clone, Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+pub enum TypeArg {
+    /// Where the TypeDef declares that an argument is a [TypeParam::Type]
+    Type(SimpleType),
+    /// Where the TypeDef declares that an argument is a [TypeParam::ClassicType],
+    /// it'll get one of these (rather than embedding inside a Type)
+    ClassicType(ClassicType),
```

Comment:
So this boils down to whether the resource-implementor writes
```
let TypeArg::ClassicType(elem_type) = arg else {panic!("Illegal argument for parameter");}
```
vs
```
let TypeArg::Type(SimpleType::Classic(elem_type)) = arg else {panic!("Illegal argument for parameter!);}
```
(of course the *calling* code has already ruled out the situations that would give rise to the panic, but the type system doesn't let us say that).

However, I think it's good to keep a 1:1 correspondence between "The TypeParam that I declare" and "The TypeArg that I expect back". Rather than TypeParam::ClassicType documenting that "you'll get back a TypeArg::Type whose contents are a SimpleType::Classic".